tools/VscPreviewExtension/src/previewCore.ts

```diff
@@ -1,39 +1,32 @@
 // Copyright (c) Microsoft. All rights reserved. Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
-import { Uri, ExtensionContext } from "vscode";
+import { ExtensionContext, window } from "vscode";
 import { MarkdownDocumentContentProvider } from "./markdownDocumentContentProvider";
-import { ChildProcessHost } from "./childProcessHost"
+import { ChildProcessHost } from "./childProcessHost";
+import * as ConstVariable from "./constVariables/commonVariables";
+import { DfmService } from "./dfmService";
 
-// Create a child process(DfmRender) by "_spawn" to render a html
 export class PreviewCore extends ChildProcessHost {
-    public isFirstTime: boolean;
     public provider: MarkdownDocumentContentProvider;
 
-    public callDfm(uri: Uri) {
-        this._documentUri = uri;
-        if (this.isFirstTime) {
-            // In the first time, if wait for the timeout, activeTextEditor will be the preview window.
-            this.isFirstTime = false;
-            this.sendMessage();
-        } else if (!this._waiting) {
-            this._waiting = true;
-            setTimeout(() => {
-                this._waiting = false;
-                this.sendMessage();
-            }, 300);
-        }
-    }
-
     protected initializeProvider(context: ExtensionContext) {
         this.provider = new MarkdownDocumentContentProvider(context);
     }
 
-    protected writeToStdin(rootPath: string, filePath: string, numOfRow: number, docContent: string) {
-        this.provider.fileName = filePath;
-        this._spawn.stdin.write(this.appendWrap("dfmmarkup"));
-        this._spawn.stdin.write(this.appendWrap(rootPath));
-        this._spawn.stdin.write(this.appendWrap(filePath));
-        this._spawn.stdin.write(this.appendWrap(numOfRow));
-        this._spawn.stdin.write(this.appendWrap(docContent));
+    protected async sendHttpRequestCoreAsync(rootPath: string, relativePath: string, docContent: string) {
+        this.provider.fileName = relativePath;
+        let that = this;
+        try {
+            let res = await DfmService.previewAsync(ChildProcessHost._serverPort, docContent, rootPath, relativePath);
+            that._isChildProcessStarting = false;
```

Comment:
No, think about this situation:
a change event trigger a request and get an error response(No service), then it will new a childProcess and waiting for a ready signal, before the childProcess ready, another change event coming, and it will get an error response, which will call newChildProcess function, and exactly this time, first Child Process is ready and set this status as false, which will let second process to new another child process. 

---
In reply to: [111710760](https://github.com/dotnet/docfx/pull/1547#discussion_r111710760) [](ancestors = 111710760)